pkg/agent/knapsack/knapsack.go

```diff
@@ -0,0 +1,84 @@
+package knapsack
+
+import (
+	"github.com/kolide/launcher/pkg/agent/storage"
+	"github.com/kolide/launcher/pkg/agent/types"
+	"go.etcd.io/bbolt"
+)
+
+// Knapsack is an inventory of data and useful services which are used throughout
+// launcher code and are typically valid for the lifetime of the launcher application instance.
+type Knapsack struct {
+	stores map[storage.Store]types.KVStore
+
+	// BboltDB is the underlying bbolt database.
+	// Ideally, we can eventually remove this. This is only here because some parts of the codebase
+	// like the osquery extension have a direct dependency on bbolt and need this reference.
+	// If we are able to abstract bbolt out completely in these areas, we should be able to
+	// remove this field and prevent "leaking" bbolt into places it doesn't need to.
+	BboltDB *bbolt.DB
+
+	// This struct is a work in progress, and will be iteratively added to as needs arise.
+	// Some potential future additions include:
+	// Flags
+	// Querier
+}
+
+func New(stores map[storage.Store]types.KVStore, db *bbolt.DB) *Knapsack {
+	k := &Knapsack{
+		BboltDB: db,
+		stores:  stores,
+	}
+
+	return k
+}
+
+func (k *Knapsack) getKVStore(storeType storage.Store) types.KVStore {
+	if k == nil {
+		return nil
+	}
+
+	// Ignoring ok value, this should only fail if an invalid storeType is provided
+	store, _ := k.stores[storeType]
```

Comment:
related to other comment, maybe create the store if it doesn't exist?

I guess that would mean rethinking the storage interface a bit.  Or perhaps a new interface that sits one level above store that allows for "bucket" creation. `StoreFactory` `StoreCreator` `StoreInit` etc

... just thinking out loud more than making a suggestion

Reply:
So, I think in the extension code or maybe osquery tables, there was some logic where the buckets were created upfront, and some corresponding error handling which expected buckets to exist when they were accessed later. That being said, we could change all the logic to be done either lazily or eagerly.


IMO one advantage of creating all the buckets upfront is that it makes error handling easier - any bucket creation errors can be handled independently, and later on you don't need to worry about if it exists or not. You can assume it does exists and If it doesn't exist, something is fubar.

Reply:
👍 yea, that makes sense